devlib/target_runner.py

```diff
@@ -0,0 +1,267 @@
+
+        * ``enable_kvm``: Specifies if KVM will be used as accelerator in QEMU or not.
+            Enabled by default if host architecture matches with target's for improving
+            QEMU performance.
+    :type qemu_settings: Dict
+
+    :param connection_settings: the dictionary to store connection settings
+        of :attr:`Target.connection_settings`, defaults to None.
+    :type connection_settings: Dict, optional
+
+    :param make_target: Lambda function for creating :class:`Target` based
+        object, defaults to :func:`lambda **kwargs: LinuxTarget(**kwargs)`.
+    :type make_target: func, optional
+
+    :Variable positional arguments: Forwarded to :class:`TargetRunner`.
+
+    :raises FileNotFoundError: if QEMU executable, kernel or initrd image cannot be found.
+    """
+
+    def __init__(self,
+                 qemu_settings,
+                 connection_settings=None,
+                 # pylint: disable=unnecessary-lambda
+                 make_target=lambda **kwargs: LinuxTarget(**kwargs),
+                 **args):
+        self.connection_settings = {
+            'host': '127.0.0.1',
+            'port': 8022,
+            'username': 'root',
+            'password': 'root',
+            'strict_host_check': False,
+        }
+
+        if connection_settings is not None:
+            self.connection_settings = self.connection_settings | connection_settings
+
+        qemu_args = {
+            'kernel_image': '',
+            'arch': 'aarch64',
+            'cpu_type': 'cortex-a72',
+            'initrd_image': '',
+            'mem_size': 512,
+            'num_cores': 2,
+            'num_threads': 2,
+            'cmdline': 'console=ttyAMA0',
+            'enable_kvm': True,
+        }
+
+        qemu_args = qemu_args | qemu_settings
+
+        qemu_executable = f'qemu-system-{qemu_args["arch"]}'
+        qemu_path = which(qemu_executable)
+        if qemu_path is None:
+            raise FileNotFoundError(f'Cannot find {qemu_executable} executable!')
+
+        if not os.path.exists(qemu_args["kernel_image"]):
+            raise FileNotFoundError(f'{qemu_args["kernel_image"]} does not exist!')
+
+        # pylint: disable=consider-using-f-string
+        qemu_cmd = '''\
+{} -kernel {} -append "{}" -m {} -smp cores={},threads={} -netdev user,id=net0,hostfwd=tcp::{}-:22 \
+-device virtio-net-pci,netdev=net0 --nographic\
+'''.format(
+            qemu_path,
+            qemu_args["kernel_image"],
+            qemu_args["cmdline"],
+            qemu_args["mem_size"],
+            qemu_args["num_cores"],
+            qemu_args["num_threads"],
+            self.connection_settings["port"],
+        )
+
+        if qemu_args["initrd_image"]:
+            if not os.path.exists(qemu_args["initrd_image"]):
+                raise FileNotFoundError(f'{qemu_args["initrd_image"]} does not exist!')
+
+            qemu_cmd += f' -initrd {qemu_args["initrd_image"]}'
+
+        if qemu_args["arch"] == machine():
```

Comment:
Are we sure that whatever Python returns in platform.machine() actually agrees with qemu arch names ? This is typically a can of worm, where half the world uses `arm64`, the other half `aarch64` and I'm not even talking about 32bits arm. x86 is usually ok, but even there you can end up with `x86-64` vs `x86_64`

Reply:
Relaxed the check by `machine().startswith('x86')`.

Reply:
Is it correct though ? I know that for now arm64 hosts are not _really_ supported in LISA, but devlib might have a different standing there, and we also don't want to prevent it from working the day the rest (e.g. adb) is actually ready.

Reply:
This is the new version of the code in #677:
```Python
        if qemu_args['arch'].startswith('x86') and machine().startswith('x86'):
            if qemu_args["enable_kvm"]:
                qemu_cmd += ' --enable-kvm'
        else:
            qemu_cmd += f' -machine virt -cpu {qemu_args["cpu_type"]}'
```
I want to enable KVM only for x86 and specify `-machine virt -cpu ...` only if the target machine/guest is Arm.
Will this hurt devlib's current arm64 support?

Reply:
> I want to enable KVM only for x86

Is KVM not working on aarch64 ?

> specify -machine virt -cpu ... only if the target machine/guest is Arm

What does `-machine virt` do that we don't want done on x86 ?

Reply:
> Is KVM not working on aarch64 ?

Works for sure, but there are some restrictions: https://stackoverflow.com/a/71220201.

> What does -machine virt do that we don't want done on x86 ?

`-machine virt` is unsupported for x86_64. `qemu-system-x86_64 -machine help | grep -i virt` returns no match.

Reply:
I had a look at those restrictions and they don't seem to be too problematic. No-one is running an arm 32bits host these days for any CI system AFAIK, and `-cpu host -machine gic-version=host` does not sound too bad considered the speed benefits (unless this has a consequence I can't imagine).

 >-machine virt is unsupported for x86_64
 
Fair enough, so that's really attached to x86 target itself and not assuming the host is x86.

Make sure that in both cases, you add a comment stating the reason why this code takes those decisions so it can be revisited easily in the future (e.g. one day x86_64 might support virt).

Reply:
`-cpu host` and `gic-version` options are not supported for `qemu-system-aarch64` on an `x86_64` host.
Also https://wiki.ubuntu.com/ARM64/QEMU says that it's not possible to use KVM accelerator for a different architecture (e.g., x86_64 host, aarch64 guest) which makes sense to me.

Reply:
Made a small improvement regarding the host/guest checks. If we don't have any open comments left here, let's continue reviewing on #677 please.